rust/datafusion/src/optimizer/constant_folding.rs

```diff
@@ -0,0 +1,834 @@
+                .collect::<Result<_>>()?,
+        },
+        Expr::ScalarUDF { fun, args } => Expr::ScalarUDF {
+            fun: fun.clone(),
+            args: args
+                .iter()
+                .map(|e| optimize_expr(e, schemas))
+                .collect::<Result<_>>()?,
+        },
+        Expr::AggregateFunction {
+            fun,
+            args,
+            distinct,
+        } => Expr::AggregateFunction {
+            fun: fun.clone(),
+            args: args
+                .iter()
+                .map(|e| optimize_expr(e, schemas))
+                .collect::<Result<_>>()?,
+            distinct: *distinct,
+        },
+        Expr::AggregateUDF { fun, args } => Expr::AggregateUDF {
+            fun: fun.clone(),
+            args: args
+                .iter()
+                .map(|e| optimize_expr(e, schemas))
+                .collect::<Result<_>>()?,
+        },
+        Expr::Sort {
+            expr,
+            asc,
+            nulls_first,
+        } => Expr::Sort {
+            expr: Box::new(optimize_expr(expr, schemas)?),
+            asc: *asc,
+            nulls_first: *nulls_first,
+        },
+        Expr::Column { .. }
+        | Expr::ScalarVariable { .. }
+        | Expr::Literal { .. }
+        | Expr::Wildcard => e.clone(),
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::logical_plan::{
+        col, lit, max, min, DFField, DFSchema, LogicalPlanBuilder,
+    };
+
+    use arrow::datatypes::*;
+
+    fn test_table_scan() -> Result<LogicalPlan> {
+        let schema = Schema::new(vec![
+            Field::new("a", DataType::Boolean, false),
+            Field::new("b", DataType::Boolean, false),
+            Field::new("c", DataType::Boolean, false),
+            Field::new("d", DataType::UInt32, false),
+        ]);
+        LogicalPlanBuilder::scan_empty("test", &schema, None)?.build()
+    }
+
+    fn expr_test_schema() -> DFSchemaRef {
+        Arc::new(
+            DFSchema::new(vec![
+                DFField::new(None, "c1", DataType::Utf8, true),
+                DFField::new(None, "c2", DataType::Boolean, true),
+            ])
+            .unwrap(),
+        )
+    }
+
+    #[test]
+    fn optimize_expr_not_not() -> Result<()> {
+        let schema = expr_test_schema();
+        assert_eq!(
+            optimize_expr(
+                &Expr::Not(Box::new(Expr::Not(Box::new(Expr::Not(Box::new(col(
+                    "c2"
+                ))))))),
+                &[&schema],
+            )?,
+            col("c2").not(),
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn optimize_expr_null_comparision() -> Result<()> {
+        let schema = expr_test_schema();
+
+        assert_eq!(
+            optimize_expr(
+                &Expr::BinaryExpr {
+                    left: Box::new(lit(true)),
+                    op: Operator::Eq,
+                    right: Box::new(Expr::Literal(ScalarValue::Boolean(None))),
```

Comment:
I am personally a fan of `col(lit(true)).eq(lit(ScalarValue::Boolean(None))))` type expression